### ISSUE TITLE
Corsor Point Problem 

### ISSUE THREAD

**Issue by codesbangla** [spckio/spck-issues#570]
I try to write a blog with Bengali words but the position of the words does not show in the right place.  Usually, adding or removing other words by clicking where the words should be is changing the words in other places.  This editor is optimized for all languages ​​besides English.  This is my favorite editor.

**Comment by spck-io** [comment]
Thanks for the bug report. There is a new mode for editing .txt files. This editor is optimized for code editing and there are some behaviors outside my control. Can you attach some screenshots indicating what you mean by your description.

**Comment by codesbangla** [comment]
![IMG_20221223_101124](https://user-images.githubusercontent.com/100434338/209269478-22fe6c0d-cee6-4506-871f-4012e82498f8.jpg)
![IMG_20221223_101106](https://user-images.githubusercontent.com/100434338/209269495-1eae5656-b8c1-4027-8114-1a4d69a9249d.jpg)
Issues: Cursor is not in correct position specially when we write something without English.

**Comment by spck-io** [comment]
I see, are you using the default font "Noto Mono"? This is an issue when characters outside the font character sets appear, as the editor has a pre-calculated width for each character and if the character is not included in the font it does not fit this width. The default Noto Mono has the majority of characters but maybe it's missing Bengali.

**Comment by codesbangla** [comment]
What can I do now?  This is my most favorite editor 😭😭😭

**Comment by codesbangla** [comment]
![New Project 24  001B595](https://user-images.githubusercontent.com/100434338/209273539-5e7c6485-61dc-4eb3-bf16-0f8f43dea561.png)
Now I am using "Incosolata" font in both editors.  No problem with TrebEdit Spck is having problem. I don't think it's a font issue.

**Comment by spck-io** [comment]
Hi, I have investigated this issue over the last few days and determined that something can be done with the editor to support Bengali. However, it involves rewriting a lot of the core editor functions and may slow down the editor performance for really large files.

I have tried to incorporate some optimization, but I'm current stuck with a particular aspect of Bengali characters/glyphs. I noticed that for "ঃ" a single character has 1 width, but repeating the character changes it's width "ঃঃ". Is this a common case in Bengali or other languages where the size of the character depends on the characters before/after it? Can you explain in which cases this happens and how common is it? Is it ok to separate out the characters into ঃ ঃ to improve editor performance?

**Comment by spck-io** [comment]
Also, if speakers of other languages can help: Does other languages have scenarios where the character width is not fixed and depends on characters before/after it? If this is the case, the optimizations I'm doing by caching the widths of certain characters will not work for these cases.

**Comment by deejinng** [comment]
> Also, if speakers of other languages can help: Does other languages have scenarios where the character width is not fixed and depends on characters before/after it? If this is the case, the optimizations I'm doing by caching the widths of certain characters will not work for these cases.

The case is simple.

Tell them to select to select a powerline font(Fira, Menslo, and lot more) some are available in the SPCK Editor fonts selection tab.

Some characters are not supported in the Roboto/Noto Mono font.

**Comment by codesbangla** [comment]
Sorry for the late reply
"ঃ" It is a dimensionless letter of Bengali language.  It is called Bisarga. Its use has declined now.  There are total three dimensionless characters in Bengali.  It cannot be used alone but must be used behind some letters.  Such as "দুঃখ" "নিঃশ্বাস".  Using it alone is supposed to make it look like it's completely normal.

**Comment by codesbangla** [comment]
":" It is a dimensionless letter of Bengali language. It is called Bisarga.
Its use has declined now. There are total three dimensionless characters in
Bengali. It cannot be used alone but must be used behind some letters. Such
as sad breath. Using it alone is supposed to make it look like it's
completely normal.

**Comment by spck-io** [comment]
Thanks for your help. "ঃ" seems to belong to a class of unicode characters called Spacing Mark. There are more cases of special unicode characters like Non-marking space that I have added support for. Added support for variable width glyphs in the next update of Spck Editor (7.0.5). This includes variable width support for CJK, all Indic languages (including Bengali), and variable-width emojis, also fixes with RTL languages like Arabic, although the BIDI algorithm implementation is severely outdated, so mixing English with Arabic repeatedly will introduce bugs.
